Constrain abstract specs rather than concatenating strings in the "when" context manager

alalazo opened the pull request:
fixes #26656

Modifications:
- [x] Add a function to merge abstract anonymous specs without resorting to any repository lookup
- [x] Implemented the when context manager based on that function

alalazo committed:
Allow prepending constraints from the "when" context manager

fixes #26656

When a directive is constrained on a dependency of the current
package we might need to prepend, rather than append, constraints
from the "when" context manager.

This PR allows users to do that by explicitly setting a
keyword argument.

haampie commented:
I'm not a fan... why not just error whenever a ^x involved?

~I see this is tagged bugfix, but why not agree that #26656 is ambiguous and we should error when things get ambiguous or unclear in meaning.~ see below.

haampie commented:
In fact, thinking about this more, I think this PR just shows we're applying the context manager on the wrong level: it should be applied on the level of abstract `Spec`s, it should not be applied on the spec-dsl-strings.

So if I have:

```python
with when(spec_str_a):
  depends_on(x, when=spec_str_b)
```

Then the resulting when condition should be the intersection `Spec(spec_str_a) ∩ Spec(spec_str_b)`.

E.g.:

```python
with when('@3:'):
  depends_on('x', when='@:4')
```

should be equivalent to `depends_on('x', when='@3:4')`

and

```python
with when('@3: ^python@3.2:'):
  depends_on('x', when'@:4 ^python@3:')
```

should be `depends_on('x', when='3:4 ^python@3.2:')`.

So, just take the intersection of the abstract specs.

haampie reviewed (changes requested):
This should not be a string operation but a spec intersection operation. There shouldn't be any mention of string processing talk

alalazo commented:
We can definitely try using a `Spec.constrain`, let me check the implementation details.

alalazo commented:
So, constraining seems to be the right thing to do, but revelead an error with `CudaPackage` and `Cuda` because of:
```python
with when('~allow-unsupported-compiler'):
    pass
```
in the `CudaPackage` class. That should read `^cuda~allow-unsupported-compiler` but using that in core Spack triggers circular import error. The issue stems from the following facts:
1. `Spec.constrain` might need a repository to check details of the specs
2. `CudaPackage` needs to load `Cuda` if there are two or more `^cuda` to be merged as specs
3. `Cuda` needs to load `CudaPackage` since that is always imported explicitly through our prepend magic

I'll try to add a `merge_abstract_specs` function that doesn't rely on metadata from repository, since that is the function we want. Another thing to be noted is that loading a package while evaluating another package might break our directives.

alalazo committed:
Constrain spec instead of working with strings

Using the Spec.constrain method doesn't work since it might
trigger a repository lookup which could break our directives
and triggers a circular import error.

To fix that we introduce a function to merge abstract anonymous
specs, based only on package names, which does not perform any
lookup in the repository.

haampie commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -1212,3 +1212,19 @@ def test_spec_dict_hashless_dep():
+    (['+baz', '+bar'], '+baz+bar'),
+    (['@2.0:', '@:5.1', '+bar'], '@2.0:5.1 +bar'),
+    # Anonymous specs with dependencies
+    (['^mpich@3.2', '^mpich@:4.0+foo'], '^mpich@3.2 +foo'),
```
now try `mpich@3.2` combined with the range `mpich@3.2.1:3.2.10` :D

alalazo replied:
I'm not a fan of the new version scheme to be honest. I may be old fashioned but 3.1 < 3.1.1 and this:
```console
Spack version 0.16.3
Python 3.8.10, Linux x86_64
>>> import spack.version
>>> import spack.spec
>>> s = spack.spec.Spec('@3.1.1:3.1')
>>> s.satisfies('@3.1.9')
True
>>> s = spack.spec.Spec('@3.1:3.1')
>>> s.satisfies('@3.1.9')
False
```
seems highly unintuitive to me :grimacing:

trws replied:
That bit me just the other day, I can kinda understand why it works but man is it unintuitive.  Especially when `spack install llvm@13` fails to try to install `llvm@13.0.0`...

haampie replied:
@alalazo yes, that is non-intuitive, because you're hitting a bug :upside_down_face: last one should be True.

alalazo replied:
This is non intuitive because the upper bound doesn't mean the version, but it means `next(upper_bound)`, regardless of the bug. As pointed out https://github.com/spack/spack/discussions/26557#discussioncomment-1474948 the syntax currently defines implicitly a "comparison" operator for which:
```console
x << y
y << x
x != y
```
might hold at the same time. Let's continue this thread in the discussion.

haampie replied:
It's not a good idea to throw in an example with a bug to argue some underlying concept is non-intuitive...

alalazo commented:
@spackbot run pipeline

alalazo commented:
@manuelakuhn Does this work for you?

alalazo commented:
@spackbot run pipeline

manuelakuhn commented:
Yes, at least for me this concretizes correctly now without errors. Thanks a lot!

alalazo commented:
The pipeline is failing in the same way also on develop.

haampie commented on lib/spack/spack/build_systems/cuda.py:
```diff
@@ -88,7 +88,7 @@ def cuda_flags(arch_list):
     # Linux x86_64 compiler conflicts from here:
     # https://gist.github.com/ax3l/9489132
-    with when('~allow-unsupported-compilers'):
+    with when('^cuda~allow-unsupported-compilers'):
```
this makes much more sense generally now...

haampie reviewed (approved):
Looks good to me, definitely an improvement over the string processing :D
